biz.aQute.remote/src/aQute/remote/util/AgentSupervisor.java

```diff
@@ -41,15 +42,26 @@
 	}
 
 	protected void connect(Class<Agent> agent, Supervisor supervisor, String host, int port) throws Exception {
-		while (true)
+		connect(agent, supervisor, host, port, -1);
+	}
+
+	protected void connect(Class<Agent> agent, Supervisor supervisor, String host, int port, int timeout)
+			throws Exception {
+		while (timeout == -1 || timeout > 0)
 			try {
 				Socket socket = new Socket(host, port);
 				link = new Link<Supervisor,Agent>(agent, supervisor, socket);
 				this.setAgent(link);
 				link.open();
 				return;
 			} catch (ConnectException e) {
-				Thread.sleep(200);
+				if (timeout > 0) {
+					timeout -= connectWait;
```

Comment:
I think you changes this to `timeout = Math.max(timeout - connectWait, 0);` That way `0` is the floor. The you can remove the `timeout == -1` check below.